Enhance: add support for CP437 encoding

SlySven opened the pull request:
This encoding is often used within the font within the ROM of PC Display cards and - with a lot of box drawing characters - is popular in some MUDs for inline map displays.

Additionally I have also managed to include the following extra encodings that are derived from CP437:
+ CP667 - "Mazovia" Used in MSDOS for Polish.
+ CP737 - Used in MSDOS for Greek, more popular than CP869 but is missing a couple of characters.
+ CP869 - Used in MSDOS for Greek, less popular than CP737 but more complete.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: add support for CP437 encoding

This encoding is often used within the font within the ROM of PC Display
cards and - with a lot of box drawing characters - is popular in some
MUDs for inline map displays.

Additionally I have also managed to include the following extra encodings
that are derived from CP437:
+ CP667 - "Mazovia" Used in MSDOS for Polish.
+ CP737 - Used in MSDOS for Greek, more popular than CP869 but is missing
  a couple of characters.
+ CP869 - Used in MSDOS for Greek, less popular than CP737 but more
  complete.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
@SlySven this is crashing because the `mpOutOfBandDataIncomingCodec` is null when this echo is being printed from ctelnet.cpp:252:

```
qDebug().nospace() << "cTelnet::encodingChanged(" << encoding << ") INFO - Installing a codec for OOB protocols that can handle: " << mpOutOfBandDataIncomingCodec->aliases();

```

SlySven committed:
BugFix: prevent Seg. Fault when Qt has no encoder for encoding asked for

Hadn't arisen in the past but two of the new ones in the last commit are
not supported by the codecs that Qt provide in the current (5.14.2) version
of it's libraries.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Remove: our decoders for encodings not currently supported by Qt

This is the CP667 (Mazovia) and CP737 (DOS Greek) ones, the other Greek one
with more characters but supposedly less used is okay.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

thoth-medievia commented:
The new definitions in TBuffer.cpp look right, but this is a limitation in QTextCodec, which has a list of supported codecs here: https://doc.qt.io/qt-5/qtextcodec.html - you will have to create a new codec class for this.  I found a GPL2 example of someone doing that for CP437, here: https://github.com/openmoko/openmoko-svn/blob/master/src/host/qemu-neo1973/phonesim/lib/serial/qatutils.cpp - search down for "437" and you will see it.

SlySven commented:
Ah, and reading the documentation:
> `[static] QTextCodec *QTextCodec::codecForName(const QByteArray &name)`
>
> Searches all installed `QTextCodec` objects and returns the one which best matches name; the match is case-insensitive. **Returns 0 if no codec matching the name name could be found.**
>
>Note: This function is thread-safe.

The *emphasis* is mine - but it seems that with CP437 (or one of the other ones) we are getting a case where there is NOT an encoder - which ***is*** a concern...

SlySven commented:
Of course https://doc.qt.io/qt-5/qtextcodec.html mentions:
> If Qt is compiled with ICU support enabled, most codecs supported by ICU will also be available to the application.

I *believe* the Qt libraries I am using have ICU support included.

thoth-medievia commented:
That's right - there's no encoder.  Here are the one's QT supports by default.  For everything else you would have to create the class yourself.  I put a link above where someone did that for IBM437/CP437

QT Default Supported Encodings:

Big5, Big5-HKSCS, CP949, EUC-JP, EUC-KR, GB18030, HP-ROMAN8, IBM 850, IBM 866, IBM 874, ISO 2022-JP, ISO 8859-1 to 10, ISO 8859-13 to 16, Iscii-Bng, Dev, Gjr, Knd, Mlm, Ori, Pnj, Tlg, and Tml, KOI8-R, KOI8-U, Macintosh, Shift-JIS, TIS-620, TSCII, UTF-8, UTF-16, UTF-16BE, UTF-16LE, UTF-32, UTF-32BE, UTF-32LE, Windows-1250 to 1258

thoth-medievia commented:
If ICU Supports them, perhaps we're just calling them by the wrong names?  For example, I think the common name for CP437 may be IBM437, and while using QT you can use the name they use, but if relying on ICU for the codec, we may need to look up what They call it...

vadi2 commented:
The latest build is able to change the encoding fine now, try it out.

thoth-medievia commented:
It looks like we are able to change the encoding to CP437 now :)  However, the system now crashes once connected to a server with either of the new encodings active, on sending any data.

SlySven commented:
Well the aliases that Qt knows for the CP437 codec in the (WIndows 64Bit) system I have in front of me are:
* ibm-437_P100-1995
* ibm-437
* cp437
* 437
* csPC8CodePage437
* windows-437

SlySven commented:
As for still crashing - I have an idea which I will check out in the next couple of hours...

SlySven commented:
Which MUD are you using this encoding for BTW?

thoth-medievia commented:
Medievia :)

medievia.com:4000

We need this encoding primarily so that we can pass all 128 ASCII
characters through to our inline map font.

On Thu, Apr 9, 2020 at 3:05 PM Stephen Lyons <notifications@github.com>
wrote:

> Which MUD are you using this encoding for BTW?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/3579#issuecomment-611701255>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APANYRJSPPVPMIZZGX7CSXLRLYMA3ANCNFSM4MEFLS4Q>
> .
>

SlySven commented:
:woman_facepalming: 

Ah, if *your* version (or rather the one built by us on AppVeyor) does not have the same encoders as mine does then yeah - it looks as though it will explode as soon as you try to send data - there will be the warning message:
`cTelnet::encodingChanged("CP437") WARNING - Unable to locate a codec for OOB protocols that can handle: "CP437"`
but that won't stop the Application shooting itself in the foot by trying to use a pointer that points to nothing afterwards...

thoth-medievia commented:
I think we're depending on QT being compiled with -icu in order for the
non-default encoders to be supported.  The default list of supported codecs
on QT is listed above.  It might be worth mentioning that QT is planning on
discontinuing support for ICU.  I am just using the binary from your latest
build to test.

On Thu, Apr 9, 2020 at 3:13 PM Stephen Lyons <notifications@github.com>
wrote:

> 🤦‍♀️
>
> Ah, if *your* version (or rather the one built by us on AppVeyor) does
> not have the same encoders as mine does then yeah - it looks as though it
> will explode as soon as you try to send data - there will be the warning
> message:
> cTelnet::encodingChanged("CP437") WARNING - Unable to locate a codec for
> OOB protocols that can handle: "CP437"
> but that won't stop the Application shooting itself in the foot by trying
> to use a pointer that points to nothing afterwards...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/3579#issuecomment-611704830>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APANYRKE3B32DABGC3LOPNLRLYM33ANCNFSM4MEFLS4Q>
> .
>

thoth-medievia commented:
> 
> 
> As for still crashing - I have an idea which I will check out in the next couple of hours...
Did you have any luck with this?  BTW - it seems to me like the local encoding actually works.  So long as you don't send data - the received data appears to be encoded.  I can't say for sure that it's encoded PROPERLY though.

:)

SlySven committed:
Enhance: add own encoders for CP437/667/737/869 for use on Win CI builds

Converts the names of the encodings (used internally) from `QString`s to
`QByteArrays` as that is what Qt uses for it's own management of them.

Take the translated Server Encoding name out of the static and thus
immutable TBuffer::csmEncodingTable and put them in a shared
`(QMap<QByteArray, QString>) mudlet::mEncodingNameMap`

Convert the preference dialogue to store the encoder required as a
QDataRole::User data item in each item rather than using the (translated)
text itself. This is more robust and simpler to handle. It also eliminared
the need for a couple of convertors to change between what is being used
internally to identify the encoders and what a user might use as the latter
are not actually used now.

As the way we provide the multi-byte decoding of Big5 is done it turns out
that we can supply both Big5-ETen (effectively what Qt provides as Big5+)
and Big5-HKCS, just by asking Qt for the correct one. So both are now
offered.

Check for each 8-bit Extended ASCII encoder on startup and do not offer
encoders that we do not provide - this should eliminate problems between
what the coder/developer has in their Qt libraries compared to what the
CI (installer builder) uses...

Four custom codecs are included here that I have written to support
the encodings listed - which is needed for Windows CI builds which uses
only Qt and NOT IUC libraries to provide encoders. These are integrated
into the above start-up checks and they are inserted into the system only
if what they can replace is not present.

Prepend an 'm ' to the displayed codec in the preferences to make it clear
that it is a Mudlet provided one (and might be slighly different {wrong!}
compared to an offical one). The `getServerEncoding()` result will give a
`M_` prefix to the same data so that they can be distinguished.

Replace the "CP874" encoding as it was erronously being called
"ISO 8859-3 (South Europe)" in the preferences and was probably clashing
with the "ISO 8859-3" real encoding. As it happens the proper CP874,
according to Wikipedia, has three codes that required TWO unicode points
to encode. This is not something that can be supported by the
`QVector<QChar>` based sysmtem that I have developed here {at least not
without some exceptional extra code) however the encoding IS very similiar
to CP1161 {a.k.a. Windows-874} and that CAN be done in this manner!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: add some unsaved changes from last commit & fixup CodeFactor errors

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: remove an accidental leftover

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: add a header file to fix SOME CI builds which were failing

Not clear why only the Linux ones failed and for both Clang and GCC
compilers.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: replace accidently deleted default constructor declaration

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Maybe I missed it but what's the reason behind the switch to a qbytearray?

SlySven commented:
Qt itself uses that for encoder names and the only place it is needed is in the presentation to the GUI which is in the profile preferences dialogue. Internally it turned out to be cleaner to go the same way rather than to keep switching between `QByteArray`s and `QString`s/`QLatinString`s/`QStringLiteral`s...

SlySven commented:
Right it should be working now @thoth-medievia - wanna give it a try?
![image](https://user-images.githubusercontent.com/6163092/79058831-ad0e6200-7c6a-11ea-93e5-542eaae0f822.png)
Which produces the following when passed through our system - note that some Codepoints cannot easily be displayed because they are processed/have another meaning:
* 0 `x00` is the end of a string
* 9 `x09` is a horizontal tab and moves the cursor
* 10 `x0A` is carriage-return and is intercepted in processing
* 12 `x0C` is vertical tab and is intercepted in processing
* 13 `x0D` is line-feed and is intercepted in processing
* 27 `x1B` is escape and is intercepted in processing
* 127 `x7F` is delete and cannot be displayed

![image](https://user-images.githubusercontent.com/6163092/79058882-3d4ca700-7c6b-11ea-977a-50d2c20c3d04.png)

Note that that is using the Qt/ICU Codec but I have provided a replacement that can be called into play when the ICU libraries are not present/used/linked to for four particular encodings - and when they are in use you will see an 'm ' prefix in the selector (as in the first image) and they will be listed with an `M_` prefix in the outputs from the Lua commands `getServerEncoding()` and `getServerEncodings()` e.g. my Windows system running a version built on my MSYS2 setup (with the ICU libs included):
```lua
getServerEncodingsList()
{
  "ASCII",
  "UTF-8",
  "GBK",
  "GB18030",
  "BIG5",
  "BIG5-HKSCS",
  "ISO 8859-1",
  "CP437",
  "M_CP667",
  "M_CP737",
  "CP850",
  "CP866",
  "CP869",
  "ISO 8859-10",
  "ISO 8859-11",
  "ISO 8859-13",
  "ISO 8859-14",
  "ISO 8859-15",
  "ISO 8859-16",
  "ISO 8859-2",
  "ISO 8859-3",
  "ISO 8859-4",
  "ISO 8859-5",
  "ISO 8859-6",
  "ISO 8859-7",
  "ISO 8859-8",
  "ISO 8859-9",
  "KOI8-R",
  "KOI8-U",
  "MACINTOSH",
  "WINDOWS-1250",
  "WINDOWS-1251",
  "WINDOWS-1252",
  "WINDOWS-1253",
  "WINDOWS-1254",
  "WINDOWS-1255",
  "WINDOWS-1256",
  "WINDOWS-1257",
  "WINDOWS-1258"
}
```

thoth-medievia commented:
I can confirm that this is working as expected.  Thank you for your assistance in getting this working.  We're really excited to be able to begin migrating out player base to Mudlet :)

SlySven committed:
BugFix: correct some details in some Q_ASSERT_X debug tests

Without this it might prove difficult to send `\xFF` to the Game Server
without Mudlet immediately dieing from a C++ `assert` type failure, and
since that particular byte value is vital to Telnet this would I expect be
fatal if not fixed...!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:rotating_light: Warning, the code in the https://github.com/Mudlet/Mudlet/pull/3579/commits/dc4f2a85c142760b3458f353e0ce4bdded6cc7e4 commit has a significant flaw which I think it will fail as soon as it tries to send a telnet protocol command to the Game Server if it is using one of those `m` prefixed encodings (i.e. likely to be the case on the CI build here for Windows). https://github.com/Mudlet/Mudlet/pull/3579/commits/1cb9a406a0b15a5d5cea115c0cf7a7dc1d67ca61 has the fix.

And the encoding does produce nice map graphics - I have just spent the last few hours running around the newbie area of Medieva (after I worked out I had to disable the secure connection option that had been set automatically for a new Mud profile - I mean, yeah, we should encourage the use of secure connections via OpenSSL where it is practicable - but I had no idea what `... Error during SSL handshake: error:1408F10B:SSL ...` actually meant when I first got disconnected because of it! :angry: )

Anyhow after I realised I needed to `cast magic missile imp` to get enough experience to reach level 2 and thus save my progress, I got a bit sucked in - and yes, even with the existing fonts on my system it looked quite good.  If someone was packaging up a Medieva module they could include the fonts they have for their MUD and arrange for them to be loaded up with the module - I had to restart after downloading the fonts manually... :grinning:

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -242,15 +233,14 @@ const QMap<QString, QPair<QString, QVector<QChar>>> TBuffer::csmEncodingTable =
+    {"ISO 8859-11",
         //      x0/x8          x1/x9          x2/xA          x3/xB          x4/xC          x5/xD          x6/xE          x7/xF
-       {QChar(0x20AC), QChar(0xFFFD), QChar(0xFFFD), QChar(0xFFFD), QChar(0xFFFD), QChar(0x2026), QChar(0xFFFD), QChar(0xFFFD),  // 80-87
+       {QChar(0xFFFD), QChar(0xFFFD), QChar(0xFFFD), QChar(0xFFFD), QChar(0xFFFD), QChar(0xFFFD), QChar(0xFFFD), QChar(0xFFFD),  // 80-87
```
Did this data mean to change for `8859-11`?

SlySven replied:
Yeah, take a look at: https://en.wikipedia.org/wiki/ISO/IEC_8859-11#Character_set ...
0x80 to 0x9F are undefined.

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -375,9 +413,29 @@ const QMap<QString, QPair<QString, QVector<QChar>>> TBuffer::csmEncodingTable =
+        QChar(0x03BE), QChar(0x03BF), QChar(0x03C0), QChar(0x03C1), QChar(0x03C3), QChar(0x03C2), QChar(0x03C4), QChar(0x0384),  // E8-EF
+        QChar(0x00AD), QChar(0x00B1), QChar(0x03C5), QChar(0x03C6), QChar(0x03C7), QChar(0x00A7), QChar(0x03C8), QChar(0x0385),  // F0=F7
+        QChar(0x00B0), QChar(0x00A8), QChar(0x03C9), QChar(0x03CB), QChar(0x03B0), QChar(0x03CE), QChar(0x25A0), QChar(0x00A0)}},// F8-FF
+     // There was a mis-titled CP874 table here but it was faulty as the DB to DE character are each encoded as TWO QChars
```
History is better recorded in git, not comments :)

SlySven replied:
Moving it to the next commit comment.

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -1019,14 +1064,27 @@ void TBuffer::translateToPlainText(std::string& incoming, const bool isFromServe
+    // "BIG5" or "BIG5-HKSCS" (which are not in the table) encodingLookupTable
+    // will be empty otherwise the 128 values in the returned table will be used
+    // for all the text data that gets through the following ANSI code and other
+    // out-of-band data processing - doing this mean that a (fast) lookup in the
```
```suggestion
    // out-of-band data processing - doing this means that a (fast) lookup in the
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -1019,14 +1064,27 @@ void TBuffer::translateToPlainText(std::string& incoming, const bool isFromServe
+    // will be empty otherwise the 128 values in the returned table will be used
+    // for all the text data that gets through the following ANSI code and other
+    // out-of-band data processing - doing this mean that a (fast) lookup in the
+    // QVector can be done as opposed to a a repeated switch(...) and branch to
```
```suggestion
    // QVector can be done as opposed to a repeated switch(...) and branch to
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -5156,3 +5188,49 @@ int TBuffer::lengthInGraphemes(const QString& text)
+const QList<QByteArray> TBuffer::getEncodingNames()
+{
+    static QByteArrayList results;
+    if (results.isEmpty()) {
```
Could you invert this `if` to keep indentation simpler?

SlySven replied:
:thinking: Weeell... I *suppose* I could change it to:
```CPP
{
    if (!results.empty()) {
        return results;
    }

    // all the initialisation stuff we have to do the *first* time

    return results;
}
```
Which would lose *one* level of indentation - but this is not the code model where you *fail-and-bale-out* as soon as you can on error, but the *we have to do some necessary stuff the first time* one and that single level of indentation is not harmful to following the program execution flow in this situation *IMVHO*...

vadi2 replied:
I think it looks nicer that way, yeah! It improves readability.

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -5156,3 +5188,49 @@ int TBuffer::lengthInGraphemes(const QString& text)
+                if (encoding == "CP437") {
+                    // Okay to insert our replacement TTextCodex_XXXX into the
+                    // system we must instantiate them once:
+                    auto* pTTextCodec_437 = new TTextCodec_437();
```
Could you change this to be a `std::unique_ptr`? Raw pointers and `new` can be error-prone - [modern way](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-smart) is to use smart pointers. if someone else changes this code and misses a `delete`, they'd make a memory leak - so we're in a way setting up a trap for them.

SlySven replied:
NO! You must NOT `delete` the object that these pointers to `TTextCodec` instances point at - or arrange for that to happen with a fancy smart-assed pointer. The internal Qt codec managing infra-structure takes ownership of them once they are created and used and will delete them at the appropriate time itself.

vadi2 replied:
ok!

SlySven replied:
I did have a `Q_UNUSED(...)` around the pointer in the first version - as I wasn't using it then - with a sarky comment about how it was used by Qt and not to `delete` it. Guess I need to put something back in about that...

SlySven replied:
Humm, it looks like I already *had*. That is followed by:
```CPP
                    // Now that it has been instantiated, the system knows about
                    // it - indeed it takes possession of it and we must NOT
                    // delete it ourselves!
```

vadi2 commented on src/TBuffer.h:
```diff
@@ -132,12 +133,12 @@ enum TMXPMode
-    // values are a pair of human-friendly name + encoding data
-    static const QMap<QString, QPair<QString, QVector<QChar>>> csmEncodingTable;
+    // value is the encoding data.
+    // Formerly a friendly name was stored here but that was subjected to
```
History is better for Git.

SlySven replied:
I felt I had to explain what was going on - and why it had changed - and where to find the missing details now.

I will look to see if I can reword it...

vadi2 commented on src/TTextCodec.cpp:
```diff
@@ -0,0 +1,738 @@
+
+// Data for the following two types of methods taken from entry in:
+// http://www.iana.org/assignments/character-sets/character-sets.xml
+QList<QByteArray> TTextCodec_437::aliases() const
```
QVector is preferred unless we need to interact with a Qt API that only takes a QList - if that's not the case, mind changing it?

SlySven replied:
Cannot do that - that is the virtual API signature that I am working inside.

vadi2 commented on src/TTextCodec.cpp:
```diff
@@ -0,0 +1,738 @@
+// a non-null ConverterState pointer we ought to consider the first of the three
+// int state_data[3] array and if castable to a (bool) false, i.e. zero we
+// should insert a BOM if the IgnoreHeader flag is not set. If there is no state
+// we should insert the BOM for each call:
```
I don't know too much about this - but isn't using BOM a no-no?

SlySven replied:
Depends on what the caller asks for, if there is a pointed to `ConverterState` and that has the flag `QTextCodec::IgnoreHeader` set in it's `QTextCodec::ConverterState::flags` member then yes we can ignore/not produce a BOM but not otherwise...

SlySven replied:
This is one of the things that was missing from the samples of other peoples attempts at producing `QTextCodec` derived class and if someone else comes along and reuses *my* code I'd like to think they'd have a more complete example to work from. :innocent:

vadi2 committed:
Don't leak Mudlet-specific implementation detail to Lua

vadi2 reviewed (dismissed):
Looks really good!

Could we get 3-4 screenshots with the font and the new Mudlet?

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -14387,13 +14385,21 @@ int TLuaInterpreter::getServerEncodingsList(lua_State* L)
     Host& host = getHostFromLua(L);
 
+    // don't leak if we're using a Mudlet or a Qt-supplied codec to Lua
+    auto sanitizeEncoding = [] (auto encodingName) {
```
If we are doing this, do we not need the *same* sanitization for the results from the `getServerEncodingsList()` function? :confused:

SlySven commented:
In https://github.com/Mudlet/Mudlet/pull/3579/commits/86bc36a6fb620f16dff8d98b31cf4771c74bcb09 you have sanitised the output from `getServerEncodingsList()` but that means that the list of results will not match the list of encoders that the Lua `setServerEncoder(...)` can use or that `getServerEncoder()` will report. What is the most consistent way we could do this to hide the internal detail *unless* they/we really want to find out...

 :thinking: Okay, the setter can internally check to see if the original `M_` less encoding is in the list and use that, or put it on if it is there with the prefix | and the getter/list getter can always return the encoder without an `M_` unless they are called with an optional (undocumented perhaps, or we could call it a `raw` argument and say it returns the *internal* name of the encoding {which few people will need to worry about :wink:}) `true` argument

thoth-medievia commented:
TBH we should reach out to them and have them integrate this codec into
their next release.  Pop into Medievia on "Slysven", and I'll show you
around a bit.  It's quite a Mud.

On Sun, Apr 12, 2020 at 11:21 AM Stephen Lyons <notifications@github.com>
wrote:

> *@SlySven* commented on this pull request.
> ------------------------------
>
> In src/TTextCodec.cpp
> <https://github.com/Mudlet/Mudlet/pull/3579#discussion_r407212870>:
>
> > +{
>
> +    return {"IBM869", "ibm-869", "869", "cp869", "csIBM869", "ibm-869_P100-1995", "windows-869", "cp-gr"};
>
> +}
>
> +
>
> +int TTextCodec_869::mibEnum() const
>
> +{
>
> +    return 2054;
>
> +}
>
> +
>
> +// All 8-bit values are valid for conversion from CP437 to Unicode so there will
>
> +// never be any invalid characters or any left over that cannot be converted and
>
> +// need to wait for the next chunk of data in the next call. However IF we have
>
> +// a non-null ConverterState pointer we ought to consider the first of the three
>
> +// int state_data[3] array and if castable to a (bool) false, i.e. zero we
>
> +// should insert a BOM if the IgnoreHeader flag is not set. If there is no state
>
> +// we should insert the BOM for each call:
>
>
> This is one of the things that was missing from the samples of other
> peoples attempts at producing QTextCodec derived class and if someone
> else comes along and reuses *my* code I'd like to think they'd have a
> more complete example to work from. 😇
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/3579#discussion_r407212870>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APANYRLTRMOHCR726NKID3LRMHL67ANCNFSM4MEFLS4Q>
> .
>

vadi2 commented:
@SlySven makes sense, and yeah, people really should not worry or have access to M_ codecs. I don't see a good reason for it and at best it limits us on what can we do with it in the future.

SlySven commented:
> TBH we should reach out to them and have them integrate this codec into their next release.

Um, who is the "them" you mean?

>  Pop into Medievia on "Slysven", and I'll show you around a bit. It's quite a Mud.

I did, *very* late last night/this morning...! :sleepy:  As "Slysven" BTW - what Identity do you go by?

FTR Do you support any OutOfBand protocols like GMCP/MSDP? Do you have a Discord guild - our Discord RPC integration could help {and similar support could be built into other clients - see the archived Repo at: https://github.com/discord/discord-rpc } as then your Mudlet players could advertise that they are playing your MUD (rather than just showing that they are using Mudlet. :smiling_imp:)... 

@vadi2 I think I can see what to do about it...

thoth-medievia commented:
We have a handler for ATCP built into our Telnet procedures, but it's not
enabled.  I believe that's the basis for GMCP.  IIRC GMCP is just a
byte-stream that comes in under a negotiated sub-option in Telnet.  We're
not using it for anything right now, but we're open to doing something with
that channel.  There was some talk about using it to signal game sounds
maybe 15 years ago.  It's been a while since we looked at it.  My
character's name is Thoth.

By them I was thinking we could give the derived class back to QT and
suggest they integrate it into their next release.

On Sun, Apr 12, 2020 at 12:01 PM Stephen Lyons <notifications@github.com>
wrote:

> TBH we should reach out to them and have them integrate this codec into
> their next release.
>
> Um, who is the "them" you mean?
>
> Pop into Medievia on "Slysven", and I'll show you around a bit. It's quite
> a Mud.
>
> I did, *very* late last night/this morning...! 😪 As "Slysven" BTW - what
> Identity do you go by?
>
> FTR Do you support any OutOfBand protocols like GMCP/MSDP? Do you have a
> Discord guild - our Discord RPC integration could help {and similar support
> could be built into other clients - see the archived Repo at:
> https://github.com/discord/discord-rpc } as then your Mudlet players
> could advertise that they are playing your MUD (rather than just showing
> that they are using Mudlet. 😈)...
>
> @vadi2 <https://github.com/vadi2> I think I can see what to do about it...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/3579#issuecomment-612637963>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APANYRIC6KYR5FZGAXXS2TLRMHQVPANCNFSM4MEFLS4Q>
> .
>

SlySven commented:
:thinking: One of our contributors fairly recently, (late last year I believe) added a system to support sound effects (possibly being downloaded from a Server) in Mudlet - it uses either ZuggSofts' MSP or an addition to GMCP IIRC. Ah yeah it started in #3132 ...

If you do get yourselves an operating GMCP handler into the Server then you can ***automate*** operating (with the end-users' permission) the Discord RP (Rich Presence) with our integration into/use of the Discord-RPC library (Remote procedure call - yep an unfortunate near naming clash...)

TBH I do not think Qt would be interested in the CP437 codec, they are happy to make use of the ICU integration at the moment to provide it - it is just that our current Windows CI build platform process does not.

vadi2 commented:
https://github.com/g7138580/protocol_gmcp ROM snippet to add GMCP to your game. It's a little bit more than ATCP

https://wiki.mudlet.org/w/Manual:Scripting#MUD_Client_Media_Protocol - modern MSP

thoth-medievia commented:
I believe that ICU support is planned to be discontinued by QT soon.

On Sun, Apr 12, 2020 at 1:08 PM Vadim Peretokin <notifications@github.com>
wrote:

> https://github.com/g7138580/protocol_gmcp ROM snippet to add GMCP to your
> game. It's a little bit more than ATCP
>
> https://wiki.mudlet.org/w/Manual:Scripting#MUD_Client_Media_Protocol -
> modern MSP
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/3579#issuecomment-612646823>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APANYRJU7D4NNSMMGAH3UYLRMHYRLANCNFSM4MEFLS4Q>
> .
>

SlySven commented:
Have you got a more up to date sit-rep compared to https://wiki.qt.io/Locale_Support_in_Qt_5 ? I'd like to know where they are heading!

thoth-medievia commented:
Have a look at this:

https://wiki.qt.io/Qt_5_ICU

On Sun, Apr 12, 2020 at 1:49 PM Stephen Lyons <notifications@github.com>
wrote:

> Have you got a more up to date sit-rep compared to
> https://wiki.qt.io/Locale_Support_in_Qt_5 ? I'd like to know where they
> are heading!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/3579#issuecomment-612652213>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APANYRM6AQAPLZZZAMSC2KTRMH5J7ANCNFSM4MEFLS4Q>
> .
>

SlySven commented:
And that says right at the top:
"THIS PLAN HAS NOW BEEN ABANDONED"
and refers you back to: http://wiki.qt.io/Qt-5-QLocale
which redirects automatically to: https://wiki.qt.io/Locale_Support_in_Qt_5
which has at the top:
"WARNING: This page is an out of date "WORK-IN-PROGRESS" and so may not be complete or accurate." as of the most recent version dated: 09:41, 5 September 2017‎.

Hence my request for something a bit more recent...! :wink:

SlySven committed:
Revise: implement some discussed changes from Peer Reviewing

Finish off hiding the fact that we have differently named substitute
`TTextCodec`s with modified names compared to the Qt `QTextCodec`s that we
hope (but are not always rewarded with) will be present. The modified names
will only show up in the relevant Lua API functions if they are explicitly
asked for.

Removed an unwanted comment which documents a change as a result of this
PR:
"There was a mis-titled CP874 table here but it was faulty as the DB to DE
character are each encoded as TWO `QChars`  of U+0E49 to U+0E4C
respectively followed by U+FE00 - and this `QVector` based system cannot
handle them however CP1161 a.k.a. Windows-874 is the same as the
ISO 8859-11 with the Euro at 80 and 8 other changes at 85 and 91 to 97."

Also tweak some comments as pointed out in peer review.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Ah, please remember that the graphical representation of the ***control characters*** that are **not** otherwise excluded because they are intercepted in the processing are not clearly defined. They are not declared in the ISO 8859-1 encoding and all the other Extended ASCII ones are derived from that *IIRC*.  Trying to reproduce the ones shown in https://en.wikipedia.org/wiki/Code_page_437#Character_set would be nice but is beyond the current system I have developed to encode only the top 128 character codes (128 to 255).

vadi2 commented:
The internal details of the encodings are still exposed to the user and will need to be exposed in the documentation... my experience says if you do something like "here's a thing, but you're supposed to ignore this thing", people won't be better off and informed, but more confused and wondering why did they need to learn something not relevant to them.

SlySven commented:
> The internal details of the encodings are still exposed to the user and will need to be exposed in the documentation...

Do you mean from the flag to the `getServerEncodeXXXX()` pair of functions that will only pop up in the error message **if** they provide an argument which is not suitable and we are not documenting at present, and in the UI in the encoding selector.

The former we could mention in a note at the bottom in the wiki:
"To support the use of some encodings on some OSes Mudlet has to provide its own, what is called an *encoder* to stand-in for what in other circumstances would be provided by the Qt system of libraries, in some limited cases it is useful to know when that is happening and providing a boolean `true` argument to this function will cause that encoder to be identified with a `M_` prefix."

and the latter in a tooltip:
"To support the use of some encodings on some OSes Mudlet has to provide its own, what is called an *encoder* to stand-in for what in other circumstances would be provided by the Qt system of libraries, in some cases it is useful to know when that is happening and those are identified in the lists with an `m ` prefix."

If we do not do this then it will be very difficult to determine from the end user if an encoding defect is coming from bugs in the codecs ***I*** have created or something else...

vadi2 commented:
That's a fair point, but still, it's confusing information for the non-techies. You know how a well-written technical explanation still goes over people's heads - at best they don't understand it, at worst they think the client is confusing because it uses all sorts of terminology they don't understand. This is how Mudlet becomes "confusing and complicated" for them, and we can't have that. That's my half of @Mudlet/ui-review, perhaps @Kebap has a take on this as well.

But still, we do need to debug things. How about we make a special function `getMudletDebug()` that returns the name of these internal codecs actually used and other important internal information?

That way we don't 'litter' every function with a debug flag and keep it all in one tidy spot.

vadi2 committed:
Remove debug flag from encoding lists

vadi2 commented:
https://github.com/SlySven/Mudlet/pull/10 removes the debug flag and https://github.com/Mudlet/Mudlet/pull/3633 adds the function. With that this'll be all set, and we gain a new pattern for showing important debug info!

![image](https://user-images.githubusercontent.com/110988/79537863-78901100-8083-11ea-80e9-34e80a0f575d.png)

SlySven commented:
Are you okay with the `m ` prefix in the GUI selector for the encoder selection *with* a tooltip...

vadi2 commented:
I don't think we should show it there either, for the same reasons - more confusion than help 😀

thoth-medievia commented:
We've been doing some testing on Medievia - and this latest version works perfectly!  When do you expect this will be merged in?

vadi2 commented:
It's just about ready to go and will be in the next release (https://github.com/Mudlet/Mudlet/milestones).

SlySven commented:
I am on this now - got diverted into a related matter which I realise I will have to re-insert into my task queue a bit further back.  I aim to have this done in the next 24 hours.

SlySven committed:
Revise: update original PR to fit in current development branch

Revised the Lua getMudletInfo() function to handle the encoder details
being `QByteArray`s instead of `QString`s. Also revised the text display
format better so that it can be copied-and-pasted somewhere else and
actually fit on the screen - this required including some code to wrap
the text at the comma separator. I think the code is good enough but a
second opinion in case there are any corner-cases would be welcomed.

Added a short explanation about the "m " prefix in the tool-tip for the
encoder selection widget in the preferences.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Okay I have a workable setup now - including some reworking of the `getMudletInfo()` - I have more extensive plans for that but they can wait for another day... :smiling_imp: 

![image](https://user-images.githubusercontent.com/6163092/79918112-b5615c80-8423-11ea-845d-956af7a763eb.png)

Somehow though my repo has gotten out of step with the main one so I am going to have to merge in and resolve conflicts from the development branch... :man_shrugging:

SlySven committed:
Merge branch 'development' into Enhance_addSupportForCP437AndOtherEncodings

Conflicts resolved in:
* src/TBuffer.h
* src/TLuaInterpreter.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -198,7 +198,7 @@
             </property>
             <property name="toolTip">
-             <string>&lt;p&gt;If you are playing a non-English game and seeing � instead of text, or special letters like &lt;span style=&quot; font-weight:600;&quot;&gt;ñ&lt;/span&gt; aren't showing right - try changing the encoding to UTF-8 or to one suggested by your game.&lt;/p&gt;</string>
+             <string>&lt;p&gt;If you are playing a non-English game and seeing � instead of text, or special letters like &lt;span style=&quot; font-weight:600;&quot;&gt;ñ&lt;/span&gt; aren't showing right - try changing the encoding to UTF-8 or to one suggested by your game.&lt;/p&gt;&lt;p&gt;For some encodings on some Operating Systems Mudlet itself has to provide the codec needed; if that is t ...
```
I'd like this tooltip to stay as-is because the new information does not really provide any useful information for the player, just gives more text to read which to a non-techie might be confusing.

SlySven replied:
Then you will have people saying WTF does that "m" mean. Also it give the hint that if they are getting odd characters when they have a need for an encoder in that group - like the Medival people - they can appreciate that we ought to be able to fix it if they get in touch.

vadi2 replied:
That is the purpose of `getMudletInfo()` and the internal codec should not leak to the tooltip, if it is!

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -3657,26 +3657,61 @@ int TLuaInterpreter::getMudletInfo(lua_State* L)
+        knownEncodings.append(adjustEncoding(encoding));
+    }
+
+    QString encodingNames{QLatin1String(knownEncodings.join(", "))};
```
I don't think it's this PR's scope to rework how `getMudletInfo()` does the display.

SlySven replied:
The previous one had to be reworked to handle the different way the data was changed and it had some short-comings - using the `Host::postMessage(...)` actually produced a poor layout for the information (the hanging indent meant it took up more space and the colouration didn't help and there wasn't sensible line breaking for what is a long line of text - as `postMessage(...)` expects multi-line texts to be pre-broken into lines).

vadi2 replied:
A different problem unrelated to adding CP437, I think. The old layout is far more compact than the new one, and more compact = good, less text to read.

vadi2 commented:
@thoth-medievia can we get a screenshot of the CP437 encoding in action with your game's map for the release post?

thoth-medievia commented:
Here are some screencaps to show the Medievia Map Font working with CP437
on Mudlet.  I am sure you will agree it looks BEAUTIFUL!

How can we get a Medievia profile on that default connect screen?

Also, is there any way to coordinate having a Medievia banner on your
website to coincide with the launch of support for Medievia on the 05/02
release?

We have a small marketing budget available for things like that.

Regards,

Thoth



On Thu, Apr 23, 2020 at 3:30 AM Vadim Peretokin <notifications@github.com>
wrote:

> @thoth-medievia <https://github.com/thoth-medievia> can we get a
> screenshot of the CP437 encoding in action with your game's map for the
> release post?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/3579#issuecomment-618231431>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APANYRJKP5TXFICPCPP7TKDRN7VAFANCNFSM4MEFLS4Q>
> .
>

vadi2 commented:
I don't think they came through on Github.

To list your game, check out https://wiki.mudlet.org/w/Listing_Your_MUD :)

We'll mention the game anyway when we add it to the defaults list - spend it instead on promoting it to the outside world :)

thoth-medievia commented:
![mudlet-medievia-screepcap-3](https://user-images.githubusercontent.com/62970949/80146037-29fcde00-857f-11ea-9ae0-40f316bc5687.JPG)
![mudlet-medievia-screepcap-2](https://user-images.githubusercontent.com/62970949/80146038-2a957480-857f-11ea-860e-e127feeec9c1.JPG)
![mudlet-medievia-screepcap-1](https://user-images.githubusercontent.com/62970949/80146039-2b2e0b00-857f-11ea-9f13-03c0ea082276.JPG)

vadi2 commented:
That looks amazing, could you also do .png not .jpg? That negatively impacts quality

SlySven commented:
What would be useful would be a 4:1 width:height ration .png (or *maybe* .svg) logo icon (ideally with any detail in the leftmost quarter if that is relevant) along with a few paragraphs as an introduction/tease/description {for example consider the WoTMUD one - I wrote that myself!!! :blush: }. We have the connection details (medievia.com:4000) and you do NOT use SSL (do you). Obviously Medievia is the name to use for it so I think that is all that is *needed*.

\<aside>Do you have an Application Id for Discord which we can use so that Discord RPC says "Playing Medievia" and shows your Discord guild icon (if it is stored on Medievia's Discord Guild with the key/name `server-icon`) rather than it saying "Playing Mudlet" - if we asked Vadim nicely I feel sure he would upload your Icon into our Guilds' resources {though of course we are limited to a total of 150} so it could show the Medievia icon even if using the Mudlet Guild resources...\</aside>

thoth-medievia commented:
![medievia-icon](https://user-images.githubusercontent.com/62970949/80850306-2efa0700-8be9-11ea-9721-81b91107741b.png)

Here is our 4:1 ratio icon for Mudlet.  We don't use SSL, just telnet to port 4000.  We don't presently have a Discord ID, but we will get one for inclusion in the next release.

Here are our teaser paragraphs:

"
If you take the deepest and most advanced game possible, and mix it with the best chat world around, what you have is Medievia.  Imagine...

Endless player-created areas to explore for the first time.  Ships you can take out on the open seas.  You can hunt serpents, kill monsters, attack other ships, explore new islands.  We have a 4-million room expansive wilderness that you can travel through for hours in one direction, and an under ocean world that is even more vast.  We have helpful dragons that fly you high above the wilderness to far off places, and evil dragons whose lairs you can defeat with teams of dozens of other players.

We have a custom font that provides beautifully rendered overhead maps, blending the best of text gaming with graphical navigation.

You can make your fortune trading via covered wagons across the wilderness, through adventure after adventure.  You can fight for dominance in PvP action in zones, ships, wilderness, herobattles, arenas and quests.  You can create your own bloodline for future generations of heroes to carry on your name.  You can even build your own homes and castles!  This is a small sample of our features.  The complete list would scroll for pages, and is still growing.  Medievia has been continuously developed for over 25 years.

This is a game that has the intuition to track your happiness, pride, sadness, and fear while changing itself to fit your needs!  Weather, storms, wind, fire, floods, disease, even asteroids.  This is literally the most dynamic game world ever attempted.  This is Medievia.

Do you dare enter?
"
Thank you so much to the mudlet team for helping us with all of this.  The player base is very excited to be making the transition to mudlet.  :)

Warm Regards,
Thoth

vadi2 commented:
The the reqs I mentioned in https://github.com/Mudlet/Mudlet/pull/3579#issuecomment-618608314 all good to go? (sorry if the follow-up comment was misleading)
